Add JSON Schema that is v1.1 compliant

VGirol opened the pull request:
I've only added sample files in relation with the new features of the 1.1 version.
I still have to add the rules regarding the @-members and the extension members.
Coming soon...

VGirol committed:
Add JSON Schema that is v1.1 compliant

VGirol committed:
Add JSON Schema that is v1.1 compliant

auvipy commented:
great job! thanks for handling this!

VGirol committed:
Add schema to validate a request that update a resource or a relationship

VGirol committed:
Add schema to validate a request that update a resource or a relationship

VGirol commented:
According to this PR #1127, i've added a schema to validate requests that update a resource or relationship.

VGirol commented:
Actually, the JSON schema for v1.0 uses the draft-07 version.
Do we keep this version ? Or do i need to upgrade to the latest version (2020-12) ?

auvipy commented:
1.1 should be latest jsonschema compliant, like openAPI 3.1 does, IMHO.

VGirol committed:
Update schema to 2020-12 version and add @member checks

VGirol committed:
Update schema to 2020-12 version and add @member checks

VGirol commented:
Could someone tell me what is missing for this PR to be merged ?

bradjones1 commented on _schemas/1.1/schema.json:
```diff
@@ -0,0 +1,753 @@
+      "type": "object",
+      "allOf": [
+        {
+          "$ref": "#/definitions/relationshipsForbidenMemberName"
```
```suggestion
          "$ref": "#/definitions/relationshipsForbiddenMemberName"
```

bradjones1 commented on _schemas/1.1/schema.json:
```diff
@@ -0,0 +1,753 @@
+      "type": "object",
+      "allOf": [
+        {
+          "$ref": "#/definitions/relationshipsForbidenMemberName"
```
```suggestion
          "$ref": "#/definitions/relationshipsForbiddenMemberName"
```

bradjones1 commented on _schemas/1.1/schema.json:
```diff
@@ -0,0 +1,753 @@
+      },
+      "unevaluatedProperties": false
+    },
+    "relationshipsForbidenMemberName": {
```
```suggestion
    "relationshipsForbiddenMemberName": {
```

bradjones1 reviewed (changes requested):
Addressing typo found during CI integration with Drupal's dictionary. `s/forbiden/forbidden`

VGirol committed:
Fix typo

VGirol committed:
Merge branch 'schema-1.1' of https://github.com/VGirol/json-api into schema-1.1

VGirol commented:
Corrected.

auvipy commented:
@dgeb can we consider this for 1.1?

auvipy commented:
@jelhan can you take a look here please?

jelhan commented:
> @jelhan can you take a look here please?

To be honest I'm not an expert for JSON schema. But I agree that we need to improve in that regard. Thanks a lot for following up.

Personally I would feel much more comfortable with reviewing if we would have some test coverage for the schema. Like having a set of JSON:API documents, which are validated with the schema in CI.

Also it's unclear to me how we could version the schema. v1.1 is not stable yet formally. We might need at least two schemas: one for the current stable version and one for the upcoming version.

We may want to setup such an infrastructure _before_ moving forward.

jelhan commented:
Just had a look at the proposed changes in this PR and noticed that it is already addressing some points I raised above. Seems to be a great starting point! 👏

auvipy commented:
yes there is another PR for the v1.0 :) https://github.com/json-api/json-api/pull/1600

dgeb commented:
I really appreciate this work and its importance for the spec. In my opinion, an official schema is not a blocker for the v1.1 release but it would be ideal to release both simultaneously. I think we stand a good chance to finalize the schema in the same time frame that the final v1.1 RC is being reviewed, without delaying either.

auvipy commented:
> I really appreciate this work and its importance for the spec. In my opinion, an official schema is not a blocker for the v1.1 release but it would be ideal to release both simultaneously. I think we stand a good chance to finalize the schema in the same time frame that the final v1.1 RC is being reviewed, without delaying either.

thats a reasonable plan

VGirol committed:
Change all title to more neutral text.
Add meta member to all test files with informations about errors present in the document.

VGirol committed:
Update schemas to be compliant with strict mode of ajv validator

VGirol committed:
Allow for parallel relationships in JSON schema v1.1

VGirol commented:
Is there anything that is blocking this PR? :) Should I do something? Thank you

auvipy reviewed (approved):
I think we can move with it after we complete merging the schema for v1.0

auvipy commented:
in fact we can move with it in 1.1 branch as well. in case v 1.0 need more work?

VGirol commented:
We must validate and merge v1.0 before merging it in 1.1 branch because this PR uses the validation script that is introduced in the #1600 (./_schemas/scripts/validator.js).

auvipy commented:
Ok great. Lets complete the v1.0 first

bbrala commented:
v1.0 seems complete, validator is merged. Think this could move forward.

LarsStegman commented:
Any chance this could be merged ^^
